### PR TITLE
start danaswap-api

### PR DESCRIPTION
**Pull request by Geometer1729**
Adds the config utxo and tests for it (since this is almost exactly the same as in hello-world-discovery)

There's a lot of copying from the template here on the assumption that we will eventually remove the template.

- [ ] The impure tests via `nix run .#offchain:test` were run and all are passing.
- [ ] Where applicable, grammar, punctuation, and spelling within code and comments should be plausibly correct.
- [ ] The issue(s) that the work is addressing should be linked in the PR description.
- [ ] Any significant changes beyond what's clear from the linked issue(s) are documented in the PR description

### CODE REVIEW

File: offchain/danaswap-api/src/Util.purs [hunk 1]
```diff
@@ -0,0 +1,196 @@
+module Util
```
Comment (marijanp) [inline offchain/danaswap-api/src/Util.purs:1]: this module needs to go to a `ctl-utils` package in a follow-up PR

**Review by marijanp — CHANGES_REQUESTED** [review]
This is the point where I want to see a better structure and naming for modules to avoid name collisions in the future. e.g. `module Api ...` -> `module Danaswap.Api ..` analoguously the directory structure: `src/Api.purs` -> `src/Danaswap/Api.purs`.

**Review by marijanp — APPROVED** [review]

File: offchain/danaswap-api/src/DanaSwap/CborTyped.purs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+liqudityTokenMintingPolicy :: CurrencySymbol -> Contract () MintingPolicy
+liqudityTokenMintingPolicy poolId = do
+  logDebug' "creating liquidity token minting policy"
+  logDebug' $ "pool id:" <> show poolId
+  decodeCborMp CBOR.trivial
+
+configAddressValidator :: Contract () Validator
+configAddressValidator = decodeCbor CBOR.configScript
+
+simpleNft :: TransactionInput -> Contract () MintingPolicy
```
Comment (epiro-ps) [inline offchain/danaswap-api/src/DanaSwap/CborTyped.purs:51]: A doc block above this function would be awesome.

File: onchain/danaswap/DanaSwap.hs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+standardNft :: ClosedTerm (PData :--> PMintingPolicy)
+standardNft = phoistAcyclic $
+  plam $ \outRefData _ sc -> unTermCont $ do
+    outRef :: Term _ PTxOutRef <- parseData outRefData
+    let (inputs :: Term _ (PBuiltinList PTxOutRef)) =
+          pmap # pfield @"outRef"
+            #$ pfield @"inputs"
+            #$ pfield @"txInfo" # sc
+    passert "didn't spend out ref" $ pelem # outRef # inputs
```
Comment (epiro-ps) [inline onchain/danaswap/DanaSwap.hs:57]: I posted about testing this on the plutarch discord just throw a line out.

https://discord.com/channels/922576618424250418/922577570766143498/1032400379658391592

**Review by epiro-ps — CHANGES_REQUESTED** [review]
Lots of code suggestions. Looks great though.

File: onchain/danaswap/DanaSwap.hs (hunk 3 above)
Comment (Geometer1729) [reply onchain/danaswap/DanaSwap.hs:57]: Plutus simple model is interesting, I kinda forgot about that.

Comment (epiro-ps) [reply onchain/danaswap/DanaSwap.hs:57]: In follow up work, we'll need to add two levels of testing for on-chain code, with the first one being critical to verifying the logic of the protocol and second slightly less given we're already testing validators and transactions with Plutip and CTL:
- Core Protocol testing - for testing the invariant equation and the core mathematical properties of protocol actions. These tests involve simple functions that only require the values needed to produce truth values that the validator will pass on. An example of a testing a core protocol action would be the following swap test: `Given a pool with token a of A and b of B, I can swap da for db where the value of the pool is preserved.`
- Emulated Blockchain testing - for testing validators and transactions within an emulated network. An example of a test would be here would be `When minting an NFT more than once, validation fails.` It seems that [Plutus Simple Model](https://github.com/mlabs-haskell/plutus-simple-model) could be the solve here. 

@CSVdB thoughts?

**Review by epiro-ps — APPROVED** [review]
:fire: :fire: :fire:

**Review by epiro-ps — APPROVED** [review]

**Review by marijanp — APPROVED** [review]

**Pull request merged**